### ISSUE TITLE
ulimits is too low run some application

### ISSUE THREAD

**Issue by liyimeng** [rancher/k3os#87]
When delpoying elastic helm chart, I got this error 
[1]: max file descriptors [4096] for elasticsearch process is too low, increase to at least [65535]

No matter how I configure ulimits inside the container, the error persistent. I guess it is host has low limits numbers.
is there any possible way to configure the  /etc/security/limits.conf?

**Comment by runningman84** [comment]
You could run some daemon set to configure higher values

**Comment by liyimeng** [comment]
Thanks @runningman84! I have tried to configure /etc/sysctl.conf and /etc/security/limits.conf, I don't see it is helping. May you elaborate a bit more details how this would be done?

**Comment by runningman84** [comment]
This could be an example:
https://github.com/DanielBerman/k8s_logging_files/blob/master/logging_elk/daemonset-startup.yaml

**Comment by liyimeng** [comment]
Thanks @runningman84  The official elastic helm chart has something similar to this. But my point is that even I configure ulimit on hosts, my elastic container could not see that is taken into effect.

**Comment by ntim** [comment]
The tutorial here shows how to set ulimit with an init container:

https://portworx.com/run-ha-elasticsearch-elk-azure-kubernetes-service/

**Comment by antibios** [comment]
I've tried to set the ulimit with an init container and it does not work.  Specifically: 
`      initContainers:
      - image: alpine:3.6
        command: ["sh", "-c", "ulimit", "-n", "65535"]
        name: elasticsearch-dev-init-ulimit
        securityContext:
          privileged: true`

Trying to run up an ELK stack leaves me with the following error:
ERROR: [1] bootstrap checks failed
[1]: max file descriptors [4096] for elasticsearch process is too low, increase to at least [65535]

Is there no way in the OS to set the default ulimit settings?  Or perhaps with containerd?

**Comment by st33v** [comment]
Yes, I'm having the same issue. The init containers work and I can shell into the elasticsearch container and verify that max-files is set correctly but elasticsearch doesn't seem to see it that way and I'm also getting the error `[1]: max file descriptors [4096] for elasticsearch process is too low, increase to at least [65535]`. 

This is k3s on k3os using the elk chart from rancher labs.

**Comment by idjohnson** [comment]
Confirming same thing.  RancherOS has sufficient limits to do large java pods, eg.
[rancher@rancher ~]$ ulimit -Hn
1000000

However, regardless of how i try and set the limits, it's stuck at Hard 4096 and soft 1024.

I tried the link from ntim above, creating a simple busybox to test:
```
linuxkit-54bf6465f26d [~]$ cat t.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container
    image: busybox:1.28
    command: ['sh', '-c', 'echo The app is running! && sleep 36000']
  initContainers:
  - name: increase-fd-ulimit
    image: busybox
    command: ["sh", "-c", "ulimit -n 65536"]
    securityContext:
      privileged: true
```

but that too shows the same fixed limits:
```
linuxkit-54bf6465f26d [~]$ kubectl exec -it myapp-pod -- /bin/sh
/ # ulimit -Hn
4096
/ # ulimit -Sn
1024
/ # whoami
root
/ # exit
```

the cloud init tries to set it in every way i can think:
```
ssh_authorized_keys:
- ssh-rsa AAAA---mypublickey----GeZc8kV
- github:idjohnson
write_files:
- encoding: ""
  content: |-
    fs.file-max = 65000
  owner: root
  path: /etc/sysctl.d/60-file-max.conf
  permissions: '0777'

k3os:
  sysctl:
    kernel.printk: 4 4 1 7
    fs.file-max: 65000
  data_sources:
  - aws
  - cdrom
  modules:
  - kvm
  - nvme
  dns_nameservers:
  - 8.8.8.8
  - 1.1.1.1
  ntp_servers:
  - 0.us.pool.ntp.org
  - 1.us.pool.ntp.org
  password: mypassword

rancher:
  sysctl:
    kernel.printk: 4 4 1 7
    fs.file-max: 65000
  password: mypassword
  autologin: tty1
```

Last note, i can confirm the sysctl file was written... just doesnt seem to take effect:
```
linuxkit-54bf6465f26d [~]$ cat /etc/sysctl.d/60-file-max.conf 
fs.file-max = 65000
linuxkit-54bf6465f26d [~]$ ulimit -Hn
4096
linuxkit-54bf6465f26d [~]$ sudo su -
linuxkit-54bf6465f26d [~]# ulimit -Hn
4096
```

**Comment by dweomer** [comment]
`k3os.sysctl` values are strings, be sure to correctly quote them, e.g.:
```
k3os:
  sysctl:
    kernel.printk: "4 4 1 7"
    fs.file-max: "65000"
```

**Comment by dweomer** [comment]
also, see https://github.com/kubernetes/kubernetes/issues/3595#issuecomment-450984220

**Comment by antibios** [comment]
I'm struggling to understand why k3os doesn't apply the /etc/security/limits.conf or at least allow the addition of user limits values to /var/lib/rancher/k3os/config.yaml

I attempted to write to /etc/security/limits.conf on startup.  I attempted to set ulimit on start up;
run_cmd:
- "ulimit -n 65535 && ulimit -u 4096"

neither of these seem to work.

**Comment by BadgerOps** [comment]
Hey, I'm experiencing this same issue - I am under the impression that we should _not_ attempt to edit `/k3os/system/config.yaml` but rather add new values to `/var/lib/rancher/k3os/config.yaml` and I ASSumed that adding the new values to the `/var/lib/rancher/k3os/config.yaml` would supersede the `/k3os/system/config.yaml` file - but I do not see this in practice. 

```
diff /k3os/system/config.yaml /var/lib/rancher/k3os/config.yaml
--- /k3os/system/config.yaml
+++ /var/lib/rancher/k3os/config.yaml
@@ -14,3 +14,6 @@
   token: <redacted_token>
   k3s_args:
   - agent
+  sysctl:
+    kernel.printk: "4 4 1 7"
+    fs.file-max: "65000"
```
after adding those sysctl lines and rebooting:
```
 ulimit -Hn
4096
```

**Comment by antibios** [comment]
My limited understanding is that ulimit is the user default limit not the actual system limit.  So again (if I'm not wrong about that), we need to set the security limit for the user/process, not only the system level.

Typical OS apply the settings in  /etc/security/limits.conf to set the user processes limits.  I don't see how we are supposed to do this in k3os.

> master-vm101 [~]$ cat /proc/sys/fs/file-max
> 500000
> master-vm101 [~]$ ulimit -Hn
> 4096

**Comment by dweomer** [comment]
- `/etc/security/limits.conf` is used by PAM to apply limits to user shells. PAM is not distributed with k3OS.
- modifying limits in an init container with `ulimit` will do nothing for your pod because `ulimit` is a shell builtin that modifies the limits for that shell process
- if you need higher limits than what k3s is giving you in your container then you will need to adjust them in your container which will need to be privileged and have the SYS_RESOURCE capability
- please see https://github.com/kubernetes/kubernetes/issues/3595 which remains unaddressed

**Comment by BadgerOps** [comment]
That makes sense, I was following along with https://www.digitalocean.com/community/tutorials/how-to-set-up-an-elasticsearch-fluentd-and-kibana-efk-logging-stack-on-kubernetes which uses the initcontainer in the statefulSet - I definitely didn't understand how that process was working.

I'll experiment with `SYS_RESOURCE` - thank you for the explanation!

**Comment by idjohnson** [comment]
Fwiw, we gave up on k3os.  The limit was not something we could work around.  We had no intention of rebuilding elastic containers or any of the other java based containers in our stack.  I verified k3s on ubuntu 18.04 as well as WSL2/Win 10 was fine when i modified the host OS (ubuntu in both cases) to have higher limits so the ulimit issue is specific to k3os.  In the end, we decided to go with RancherOS and RKE.  If k3os comes up with a way to sort this out, we'll try again.
